src/topics/TopicListScreen.js

```diff
@@ -1,67 +1,52 @@
 /* @flow strict-local */
 
-import React, { PureComponent } from 'react';
+import React, { useState, useCallback, useEffect } from 'react';
+import type { Node } from 'react';
 
 import type { RouteProp } from '../react-navigation';
 import type { AppNavigationProp } from '../nav/AppNavigator';
-import type { Dispatch, Stream, TopicExtended } from '../types';
-import { connect } from '../react-redux';
+import { useSelector, useDispatch } from '../react-redux';
 import { Screen } from '../common';
 import { topicNarrow } from '../utils/narrow';
 import { getTopicsForStream } from '../selectors';
 import { getStreamForId } from '../subscriptions/subscriptionSelectors';
 import TopicList from './TopicList';
 import { fetchTopics, doNarrow } from '../actions';
 
-type SelectorProps = $ReadOnly<{|
-  stream: Stream,
-  topics: ?(TopicExtended[]),
-|}>;
-
 type Props = $ReadOnly<{|
   navigation: AppNavigationProp<'topic-list'>,
   route: RouteProp<'topic-list', {| streamId: number |}>,
-
-  dispatch: Dispatch,
-  ...SelectorProps,
 |}>;
 
-type State = {|
-  filter: string,
-|};
-
-class TopicListScreen extends PureComponent<Props, State> {
-  state = {
-    filter: '',
-  };
-
-  componentDidMount() {
-    const { dispatch, stream } = this.props;
+export default function TopicListScreen(props: Props): Node {
+  const dispatch = useDispatch();
+  const stream = useSelector(state => getStreamForId(state, props.route.params.streamId));
+  const topics = useSelector(state => getTopicsForStream(state, props.route.params.streamId));
+
+  const [filter, setFilter] = useState<string>('');
+
+  const handlePress = useCallback(
+    (streamObj: string, topic: string) => {
+      // TODO: use Suspense, when available, in case of race conditions when
+      // fetching on changes to `stream`:
+      //   https://github.com/zulip/zulip-mobile/pull/4906#discussion_r673557179
+      //   https://github.com/facebook/react/issues/14326#issuecomment-441680293
+      // Currently `stream` doesn't change, so this ends up being OK.
+      dispatch(doNarrow(topicNarrow(stream.name, topic)));
+    },
```

Comment:
Hmm, I think this is actually fine.

The concern from those threads would apply here if we were taking `stream`, and then doing something with it that can take time, like fetching some data… and then we took the result and applied it to what we show in this component, but by the time we do so, `stream` may have changed so that the data we got may be an answer to an out-of-date question. We could be showing a new input value, and the output for an old input value, at the same time, so that what's on the screen is inconsistent and misleading.

But here we're not doing that. We take `stream`, and immediately use it to go initiate a navigation, and that's it.

When we do that, the important thing is just that the `stream` we use is the one the user intended -- which basically means the one that was used for rendering the UI that the user acted on.

(There is a gap in that logic, which is that the value the user actually intended to refer to is the one that was used to render the UI a few frames ago, like 30-100ms, because it takes the human eye and brain and finger that long to act. There isn't a perfect general solution to that issue, but it's an excellent reason to try to avoid jerking the UI around more than necessary, and especially right after first showing some UI, when the user may be most likely to try to interact with it.)

Anyway, so we want the `stream` that was used in rendering the UI the user acted on.

And that is exactly the `stream` that we had when we last rendered this component, because that's what gets passed down to `TopicList`. So using `stream` here is correct.

Reply:
One angle on the data-fetching race described at that upstream React thread is that we're basically completely avoiding it here by an indirection through Redux:

* The UI doesn't get the data from this component's state (which would be for a particular stream, opening up the possibility of a race where what stream that should be changes and the data doesn't get updated in sync). Instead it gets it from the global Redux state, where it looks up the data corresponding to the desired stream.
* When we fetch, we don't fetch the data into this component's state, or anything else that's about some particular stream that might be identified implicitly somewhere else. Instead we fetch it into Redux, storing it filed under whatever stream we fetched data for.

Because the data is stored in a place (our global Redux store) where it's always keyed by the stream the data corresponds to, and because the UI is always told just the stream and goes and looks up the corresponding data there, we avoid the possibility of this kind of race where the data shown in the UI isn't in sync with which stream it's supposed to be about.

Reply:
Thanks, that makes sense!